core/modules/replica/HttpIngestModule.cc

```diff
@@ -86,6 +86,43 @@ string leastLoadedWorker(DatabaseServices::Ptr const& databaseServices,
     return worker;
 }
 
+/**
+ * The optimized version of the previously defined function will
+ * re-use and update the transient replica disposition cache when
+ * selecting a candidate worker. Each worker's entry in the cache
+ * will be populated from the database first time 
+ *
+ * @param worker2replicasCache  the cache of replica disposition for workers
+ * @param databaseServices  the service used for pulling the initial replica
+ *   disposition counters for workers.
+ * @param workers  a collection of workers to be evaluated as candidates.
+ * @return the name of a worker which has the least number of replicas
+ */
+template<typename COLLECTION_OF_WORKERS>
+string leastLoadedWorker(map<string,size_t>& worker2replicasCache,
+                         DatabaseServices::Ptr const& databaseServices,
+                         COLLECTION_OF_WORKERS const& workers) {
+    string worker;
+    string const noSpecificDatabase;
+    bool   const allDatabases = true;
+    size_t numReplicas = numeric_limits<size_t>::max();
+    for (auto&& candidateWorker: workers) {
+        if (worker2replicasCache.count(candidateWorker) == 0) {
+            worker2replicasCache[candidateWorker] =
```

Comment:
IIRC Scott Meyers discourages use of the [] operator with maps. In this case I think you'd be better off using something like `emplace` and keep and use the returned iterator instead of performing so many map lookups.

Reply:
you could also skip the call to `count`; `emplace` (or `insert`) will _not_ create a new key if one already exists and in either case will return an iterator to the item at the key, which you can then carry on using.
...It looks like you may want to avoid calling `databaseServices->numWorkerReplicas` for keys that already exist in the map tho, in which case I'd call `find` and then if the returned `itr==end()` then perform the insert, and use the returned itr for the rest of the calls...

Reply:
Exactly! The goal of this optimization was to avoid making expensive (and unnecessary) database queries on each invocation of the function. As it was discovered in a large scale setup of the `lsp-int` at NCSA, it takes about `0.1 sec` to fetch the worker-specific info from MySQL at a presence of ~1 million replicas (across all catalogs). This resulted in the poor performance of the chunk allocation mechanism during the ingests:
```
0.1 sec * 30 workers * 150,000 chunks = 3 sec/chunk * 150,000 chunks = 450,000 seconds = many hours
```
The new implementation does it all within less than 9 minutes.

As far as your suggestion to move to the "iterator-based" implementation then I have a conflicting feeling about it. I must admit, I had the very same thought (as you) before I implemented the function. Then I ruled iterators out because the map is rather tiny - it has as many elements as many workers are in a Qserv setup, which is on a scale of 100. On that scale any lookup would take a few microseconds at most, which would be 10,000 - 100,000 faster compared with the database query. So, the performance of the lookup operations is not an issue here. It's obvious that iterators won't make any measurable speedup here. What really discouraged me initially from using iterators with maps (unless I have a good case for this) is an added complexity to the code. The iterator-based implementations becomes harder to read and comprehend.

Having said that, I will still look if I will find a good way of expression the algorithm's intent with iterators.